Tests/ReactiveSwiftTests/ActionSpec.swift

```diff
@@ -114,6 +114,21 @@ class ActionSpec: QuickSpec {
 				expect(action.isExecuting.value) == false
 			}
 
+			it("should not deadlock") {
+				final class ViewModel {
+					let action2 = Action<(), (), NoError> { SignalProducer(value: ()) }
+				}
+
+				let action1 = Action<(), ViewModel, NoError> { SignalProducer(value: ViewModel()) }
+
+				action1.values
+					.flatMap(.latest) { viewModel in viewModel.action2.values.map { _ in viewModel } }
```

Comment:
The deadlock happens when the observer disposable releases the closure `{ _ in viewModel }` here. The mapped signal's `updateLock` is not released while the `viewModel` is released, [causing a deadlock when the terminal event is eventually propagated to the mapped signal](https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/Signal.swift#L157).